File: module_4/python/calculate.py
"""
Title: Basic calculator

Problem:
    Implement a basic calculator to evaluate a simple expression string.

    The expression string may contain open ( and closing parentheses ), the
    plus + or minus sign -, non-negative integers and empty spaces .

Execution: python calculate.py
"""
import unittest


def evaluate_expr(stack):
    """Evaluate expression as stack."""
    res = stack.pop() if stack else 0
    # Evaluate the expression till we get corresponding ')'
    while stack and stack[-1] != ')':
        sign = stack.pop()
        if sign == '+':
            res += stack.pop()
        else:
            res -= stack.pop()
    return res


def calculate(s: str) -> int:
    """Interpret string and calculate."""
    stack = []
    n, operand = 0, 0

    for i in range(len(s) - 1, -1, -1):
        ch = s[i]

        if ch.isdigit():

            # Forming the operand - in reverse order.
            operand = (10**n * int(ch)) + operand
            n += 1

        elif ch != " ":
            if n:
                # Save the operand on the stack
                # As we encounter some non-digit.
                stack.append(operand)
                n, operand = 0, 0

            if ch == '(':
                res = evaluate_expr(stack)
                stack.pop()

                # Append the evaluated result to the stack.  This result could
                # be of a sub-expression within the parenthesis.
                stack.append(res)

            # For other non-digits just push onto the stack.
            else:
                stack.append(ch)

    # Push the last operand to stack, if any.
    if n:
        stack.append(operand)

    # Evaluate any left overs in the stack.
    return evaluate_expr(stack)


# An approach using recursion instead of stacks
def calculate_recursive(s: str, start: int = 0, end: int = None) -> int:
    if end is None:
        end = len(s)

    print (s[start:end])
    sum = 0
    current_sign = '+'

    i = start
    while i < end:
        current_char = s[i]

        if current_char == ' ':
            i = i+1
            continue

        if current_char == '+' or current_char == '-':
            current_sign = current_char
            i = i+1
            continue

        next_value = 0

        if current_char.isdigit():
            next_value = int(current_char)

        if current_char == '(':
            end_paren = get_matching_paren(s, i)
            next_value = calculate_recursive(s, i+1, end_paren);
            i = end_paren

        if current_sign == '+':
            sum = sum + next_value
        else:
            sum = sum - next_value

        i = i+1
    return sum

def get_matching_paren(s: str, start: int) -> int:
    count_open = 1
    for i in range(start+1, len(s)):
        if s[i] == '(':
            count_open = count_open+1
        if s[i] == ')':
            count_open = count_open-1
        if count_open == 0:
            return i

    return -1

class TestCalculate(unittest.TestCase):
    """Unit test for calculate."""

    def test_1(self):
        """Test for 1+1 = 2."""
        self.assertEqual(calculate("1 + 1"), 2)

    def test_2(self):
        """Test for 2-1+2 = 3."""
        self.assertEqual(calculate(" 2-1 + 2 "), 3)

    def test_3(self):
        """Test for (1+(4+5+2)-3)+(6+8) = 23"""
        self.assertEqual(calculate("(1+(4+5+2)-3)+(6+8)"), 23)


if __name__ == '__main__':
    unittest.main()

File: module_4/java/Calculate.java
/*
 *   Title: Basic calculator
 *
 *   Problem:
 *      Implement a basic calculator to evaluate a simple expression
 *      string.
 *
 *      The expression string may contain open ( and closing parentheses ), the plus +
 *      or minus sign -, non-negative integers and empty spaces .
 *
 *   Execution: javac Calculate.java && java Calculate
 */
import java.util.*;


public class Calculate {
    public static int evaluateExpr(Stack<Object> stack) {

        int res = 0;

        if (!stack.empty()) {
            res = (int) stack.pop();
        }

        // Evaluate the expression till we get corresponding ')'
        while (!stack.empty() && !((char) stack.peek() == ')')) {

            char sign = (char) stack.pop();

            if (sign == '+') {
                res += (int) stack.pop();
            } else {
                res -= (int) stack.pop();
            }
        }
        return res;
    }

    public static int calculate(String s) {

        int operand = 0;
        int n = 0;
        Stack<Object> stack = new Stack<Object>();

        for (int i = s.length() - 1; i >= 0; i--) {

            char ch = s.charAt(i);

            if (Character.isDigit(ch)) {

                // Forming the operand - in reverse order.
                operand = (int) Math.pow(10, n) * (int) (ch - '0') + operand;
                n += 1;

            } else if (ch != ' ') {
                if (n != 0) {

                    // Save the operand on the stack
                    // As we encounter some non-digit.
                    stack.push(operand);
                    n = 0;
                    operand = 0;

                }
                if (ch == '(') {

                    int res = evaluateExpr(stack);
                    stack.pop();

                    // Append the evaluated result to the stack.
                    // This result could be of a sub-expression within the parenthesis.
                    stack.push(res);

                } else {
                    // For other non-digits just push onto the stack.
                    stack.push(ch);
                }
            }
        }

        //Push the last operand to stack, if any.
        if (n != 0) {
            stack.push(operand);
        }

        // Evaluate any left overs in the stack.
        return evaluateExpr(stack);
    }

    // Using recursion rather than stacks to solve this problem
    public static int calculateRecursive(String s) {
        return calculateRecursive(s, 0, s.length());
    }

    private static int calculateRecursive(String s, int start, int end) {
        int sum = 0;
        char currentSign = '+';

        for (int i = start; i < end; i++) {
            char currentChar = s.charAt(i);

            if (currentChar == ' ') continue;

            if (currentChar == '+' || currentChar == '-') {
                currentSign = currentChar;
                continue;
            }

            int nextValue = 0;

            if (Character.isDigit(currentChar)) nextValue = Integer.parseInt(String.valueOf(currentChar));
            if (currentChar == '(') {
                int endParen = getMatchingParen(s, i);
                nextValue = calculateRecursive(s, i+1, endParen);
                i = endParen;
            }

            if (currentSign == '+') sum += nextValue;
            else sum -= nextValue;
        }

        return sum;
    }

    private static int getMatchingParen(String s, int start) {
        int countOpen = 1;
        for (int i = start+1; i < s.length(); i++) {
            if (s.charAt(i) == '(') countOpen++;
            if (s.charAt(i) == ')') countOpen--;
            if (countOpen == 0) return i;
        }
        return -1;
    }


    public static void main(String[] args) {
        assert calculate("1 + 1") == 2;
        assert calculate(" 2-1 + 2 ") == 3;
        assert calculate("(1+(4+5+2)-3)+(6+8)") == 23;

        System.out.println("Passed all test cases");
    }

}
